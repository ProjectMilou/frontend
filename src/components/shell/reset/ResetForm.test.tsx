import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import * as React from 'react';
import ResetForm from './ResetForm';

beforeEach(() => {
  fetchMock.resetMocks();
});

describe('ResetForm', () => {
  const renderWindow = () => {
    const props = {
      setSuccess: jest.fn(),
    };
    return {
      ...render(<ResetForm {...props} />),
      ...props,
    };
  };

  test('should submit email', async () => {
    renderWindow();
    const input = screen.getByLabelText('shell.email') as HTMLInputElement;
    const button = screen.getByText(/shell.reset.submit/i);
    act(() => {
      fireEvent.change(input, { target: { value: 'milou@test.com' } });
    });

    fetchMock.mockResponseOnce('{ok: true}', { status: 200 });

    await waitFor(() => {
      fireEvent.click(button);
    });
    expect(window.fetch).toHaveBeenCalledWith(
      'https://api.milou.io/user/reset/forgot',
      expect.objectContaining({
        body: '{"email":"milou@test.com"}',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        method: 'POST',
      })
    );
    expect(window.fetch).toHaveBeenCalledTimes(1);
  });

  test('should show error', async () => {
    renderWindow();
    const input = screen.getByLabelText('shell.email') as HTMLInputElement;
    const button = screen.getByText(/shell.reset.submit/i);
    act(() => {
      fireEvent.change(input, { target: { value: 'milou@test.com' } });
    });

    fetchMock.mockResponseOnce('{ok: true}', { status: 400 });

    await waitFor(() => {
      fireEvent.click(button);
    });
    expect(window.fetch).toHaveBeenCalledWith(
      'https://api.milou.io/user/reset/forgot',
      expect.objectContaining({
        body: '{"email":"milou@test.com"}',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        method: 'POST',
      })
    );
    expect(window.fetch).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(screen.getByText(/error.invalidEmail/i)).toBeInTheDocument();
    });
  });

  test('should undo email error', async () => {
    renderWindow();
    const input = screen.getByLabelText('shell.email') as HTMLInputElement;
    const button = screen.getByText(/shell.reset.submit/i);

    await waitFor(() => {
      fireEvent.click(button);
    });

    await waitFor(() => {
      expect(screen.getByText(/error.invalidEmail/i)).toBeInTheDocument();
    });

    act(() => {
      fireEvent.change(input, { target: { value: 'milou@test.com' } });
    });

    expect(screen.queryByText(/error.invalidEmail/i)).not.toBeInTheDocument();
  });

  test('should display text', () => {
    const { container } = renderWindow();
    expect(container).toMatchSnapshot();
  });
});
