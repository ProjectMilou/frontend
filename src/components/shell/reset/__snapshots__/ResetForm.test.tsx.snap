// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetForm should display text 1`] = `
<div>
  <div
    class="MuiDialogContent-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
    >
      shell.reset.forgot
    </p>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style="margin: 10px 0px;"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="email"
        id="email-label"
      >
        shell.email
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="email"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        shell.reset.submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
